File: Packages/MeterKit/Sources/MeterPersistence/TransferLockout.swift
import Foundation

/// 只挡这台设备上的乱试。挡不住已经拿到文件的人离线穷举。
public struct TransferLockout: Equatable, Sendable {
    public static let freeAttemptLimit = 5

    public private(set) var failureCount: Int
    public private(set) var lockedUntil: Date?

    public init(failureCount: Int = 0, lockedUntil: Date? = nil) {
        self.failureCount = failureCount
        self.lockedUntil = lockedUntil
    }

    public var remainingFreeAttempts: Int {
        max(0, Self.freeAttemptLimit - failureCount)
    }

    public mutating func registerFailure(now: Date) {
        failureCount += 1
        let delay = Self.delay(afterFailureCount: failureCount)
        if delay > 0 {
            lockedUntil = now.addingTimeInterval(delay)
        }
    }

    public mutating func registerSuccess() {
        failureCount = 0
        lockedUntil = nil
    }

    public func isLocked(now: Date) -> Bool {
        guard let lockedUntil else { return false }
        return now < lockedUntil
    }

    public func remainingSeconds(now: Date) -> Int {
        guard let lockedUntil else { return 0 }
        return max(0, Int(ceil(lockedUntil.timeIntervalSince(now))))
    }

    /// 第 5 次起才锁，之后时间递增。这不是密码学保护。
    public static func delay(afterFailureCount count: Int) -> TimeInterval {
        switch count {
        case ...4:
            return 0
        case 5:
            return 15
        case 6:
            return 30
        case 7:
            return 60
        case 8:
            return 120
        case 9:
            return 300
        default:
            return 900
        }
    }
}

File: Android/app/src/main/kotlin/com/zhechengqi/tollcat/settings/TransferLockout.kt
package com.zhechengqi.tollcat.settings

import kotlin.math.ceil
import kotlin.math.max

/**
 * 只挡这台设备上的乱试。挡不住已经拿到文件的人离线穷举。
 * 对齐 iOS `MeterPersistence.TransferLockout`。
 */
data class TransferLockout(
    val failureCount: Int = 0,
    val lockedUntilMillis: Long? = null,
) {
    val remainingFreeAttempts: Int
        get() = max(0, FREE_ATTEMPT_LIMIT - failureCount)

    fun registerFailure(nowMillis: Long): TransferLockout {
        val nextCount = failureCount + 1
        val delayMs = delayAfterFailureCount(nextCount)
        return copy(
            failureCount = nextCount,
            lockedUntilMillis = if (delayMs > 0) nowMillis + delayMs else lockedUntilMillis,
        )
    }

    fun registerSuccess(): TransferLockout = TransferLockout()

    fun isLocked(nowMillis: Long): Boolean {
        val until = lockedUntilMillis ?: return false
        return nowMillis < until
    }

    fun remainingSeconds(nowMillis: Long): Int {
        val until = lockedUntilMillis ?: return 0
        return max(0, ceil((until - nowMillis) / 1000.0).toInt())
    }

    companion object {
        const val FREE_ATTEMPT_LIMIT = 5

        /** 第 5 次起才锁，之后时间递增。这不是密码学保护。 */
        fun delayAfterFailureCount(count: Int): Long {
            val seconds = when {
                count <= 4 -> 0L
                count == 5 -> 15L
                count == 6 -> 30L
                count == 7 -> 60L
                count == 8 -> 120L
                count == 9 -> 300L
                else -> 900L
            }
            return seconds * 1000L
        }
    }
}
